session/session.go

```diff
@@ -164,8 +164,26 @@ func (s *session) getMembufCap() int {
 func (s *session) cleanRetryInfo() {
 	if !s.sessionVars.RetryInfo.Retrying {
 		retryInfo := s.sessionVars.RetryInfo
-		for _, stmtID := range retryInfo.DroppedPreparedStmtIDs {
-			delete(s.sessionVars.PreparedStmts, stmtID)
+		if len(retryInfo.DroppedPreparedStmtIDs) > 0 {
+			planCacheEnabled := plannercore.PreparedPlanCacheEnabled()
```

Comment:
Will the `deallocate prepare stmt1` in the retry history ?

Reply:
I do some dig for it..

the result is `deallocate prepare stmt1` will be added into history, and protocol level `stmtClose` doesn't..

just as https://github.com/pingcap/tidb/pull/2473#discussion_r96139060 said, we only add exec into history, and retry reuse exec, so that PR fix the `stmtClose` delay stmt distory until retry finished....

but It seems forgot `deallocate prepare stmt1`, it seems there are a bug in `deallocate prepare stmt1` with retry in master code :rofl:  :rofl:  :rofl:  :rofl:  :rofl:

Reply:
@tiancaiamao I retrid:

- for `prepare stmt1 from 'xx'` and `deallocate prepare stmt1` will add history, and will reprepare and close for every retry, so no need lazy clean, and it's unnormal for people use prepare in text protocol, maybe is ok.
- but for binary `stmtPrepare` and `stmtClose` will NOT add to history, so need lazy cleanup at here

summary: it seems no problem.

Reply:
retry `prepare stmt1 from 'xx'`  will prepare twice, and retry `deallocate prepare stmt1` will dealloc twice?
What will happen in the plan cache ?

Reply:
after this PR will it will add plan cache twice and remove from plan cache twice...

I suddenly found there are maybe have question when a "prepare stmt from xx" in a transaction but without dealloc, retry will make many useless stmt in server, maybe handleStmtPrepare's way is better, we should use a unified way to handle this two entrances.